File: backend/routers/quiz.py
from fastapi import APIRouter, Depends
from backend.models.user_model import UserProfile
from backend.models.quiz_model import (
    QuizGenerateRequest,
    QuizGenerateResponse,
    QuizSubmitRequest,
    QuizResultResponse,
    AttemptSummary,
    QuizStatsResponse,
)
from backend.services.quiz_service import QuizService
from backend.middleware.auth_middleware import get_current_user

router = APIRouter(prefix="/quiz", tags=["quiz"])
quiz_service = QuizService()


@router.post("/generate", response_model=QuizGenerateResponse)
async def generate_quiz(
    body: QuizGenerateRequest,
    current_user: UserProfile = Depends(get_current_user),
):
    return quiz_service.generate_questions(
        current_user.id, body.topics, body.difficulty, body.total_questions
    )


@router.post("/submit", response_model=QuizResultResponse)
async def submit_quiz(
    body: QuizSubmitRequest,
    current_user: UserProfile = Depends(get_current_user),
):
    return quiz_service.submit_attempt(current_user.id, body.attempt_id, body.answers)


@router.get("/attempts", response_model=list[AttemptSummary])
async def list_attempts(current_user: UserProfile = Depends(get_current_user)):
    return quiz_service.get_user_attempts(current_user.id)


@router.get("/attempts/{attempt_id}", response_model=QuizResultResponse)
async def get_attempt(
    attempt_id: str,
    current_user: UserProfile = Depends(get_current_user),
):
    return quiz_service.get_attempt_detail(current_user.id, attempt_id)


@router.get("/stats", response_model=QuizStatsResponse)
async def get_stats(current_user: UserProfile = Depends(get_current_user)):
    return quiz_service.get_skill_stats(current_user.id)

File: frontend/lib/quiz.ts
import api from "./axios"

export interface QuizGenerateRequest {
  topics: string[]
  difficulty: string
  total_questions: number
}

export interface QuestionResponse {
  question: string
  options: string[]
  topic: string
  difficulty: string
}

export interface QuizGenerateResponse {
  attempt_id: string
  time_limit_minutes: number
  questions: QuestionResponse[]
}

export interface QuizSubmitRequest {
  attempt_id: string
  answers: number[]
}

export interface TopicScore {
  correct: number
  total: number
  percentage: number
}

export interface QuestionResult {
  question: string
  options: string[]
  correct_option: number
  selected_option: number
  is_correct: boolean
  explanation: string
  topic: string
  difficulty: string
}

export interface QuizResultResponse {
  attempt_id: string
  total_questions: number
  correct_answers: number
  score_percentage: number
  topic_scores: Record<string, TopicScore>
  skill_scores_updated: Record<string, number>
  questions: QuestionResult[]
  created_at?: string
}

export interface AttemptSummary {
  id: string
  topics: string[]
  difficulty: string
  total_questions: number
  correct_answers: number
  score_percentage: number
  created_at?: string
}

export interface QuizStatsResponse {
  skill_scores: Record<string, number>
}

export async function generateQuiz(
  data: QuizGenerateRequest
): Promise<QuizGenerateResponse> {
  const res = await api.post<QuizGenerateResponse>("/quiz/generate", data)
  return res.data
}

export async function submitQuiz(
  attempt_id: string,
  answers: number[]
): Promise<QuizResultResponse> {
  const res = await api.post<QuizResultResponse>("/quiz/submit", {
    attempt_id,
    answers,
  })
  return res.data
}

export async function listAttempts(): Promise<AttemptSummary[]> {
  const res = await api.get<AttemptSummary[]>("/quiz/attempts")
  return res.data
}

export async function getAttempt(
  id: string
): Promise<QuizResultResponse> {
  const res = await api.get<QuizResultResponse>(`/quiz/attempts/${id}`)
  return res.data
}

export async function getStats(): Promise<QuizStatsResponse> {
  const res = await api.get<QuizStatsResponse>("/quiz/stats")
  return res.data
}
